algorithms/linfa-nn/src/balltree.rs

```diff
@@ -0,0 +1,331 @@
+use std::{cmp::Reverse, collections::BinaryHeap, marker::PhantomData};
+
+use linfa::Float;
+use ndarray::{Array1, Array2};
+use noisy_float::{checkers::FiniteChecker, NoisyFloat};
+
+use crate::{
+    distance::Distance,
+    heap_elem::{MaxHeapElem, MinHeapElem},
+    BuildError, NearestNeighbour, NearestNeighbourIndex, NnError, Point,
+};
+
+// Partition the points using median value
+fn partition<F: Float>(mut points: Vec<Point<F>>) -> (Vec<Point<F>>, Point<F>, Vec<Point<F>>) {
+    debug_assert!(points.len() >= 2);
+
+    // Spread of a dimension is measured using range, which is suceptible to skew. It may be better
+    // to use STD or variance.
+    let max_spread_dim = (0..points[0].len())
+        .map(|dim| {
+            // Find the range of each dimension
+            let it = points
+                .iter()
+                .map(|p| NoisyFloat::<_, FiniteChecker>::new(p[dim]));
+            // May be faster if we can compute min and max with the same iterator, but compiler might
+            // have optimized for that
+            let max = it.clone().max().expect("partitioned empty vec");
+            let min = it.min().expect("partitioned empty vec");
+            (dim, max - min)
+        })
+        .max_by_key(|&(_, range)| range)
+        .expect("vec has no dimensions")
+        .0;
+
+    let mid = points.len() / 2;
+    // Compute median on the chosen dimension in linear time
+    let median = order_stat::kth_by(&mut points, mid, |p1, p2| {
+        p1[max_spread_dim]
+            .partial_cmp(&p2[max_spread_dim])
+            .expect("NaN in data")
+    })
+    .clone();
+
+    let (mut left, mut right): (Vec<_>, Vec<_>) = points
+        .into_iter()
+        .partition(|pt| pt[max_spread_dim] < median[max_spread_dim]);
+    // We can get an empty left partition with degenerate data where all points are equal and
+    // gathered in the right partition.  This ensures that the larger partition will always shrink,
+    // guaranteeing algorithm termination.
+    if left.is_empty() {
+        left.push(right.pop().unwrap());
+    }
+    (left, median, right)
+}
+
+// Calculates radius of a bounding sphere
+fn calc_radius<'a, F: Float, D: Distance<F>>(
+    points: impl Iterator<Item = Point<'a, F>>,
+    center: Point<F>,
+    dist_fn: &D,
+) -> F {
+    let r_rad = points
+        .map(|pt| NoisyFloat::<_, FiniteChecker>::new(dist_fn.rdistance(pt.clone(), center)))
```

Comment:
can't you accept a reference for your distance function?

Reply:
The distance function already accepts `ArrayView`, which is already a reference type, so I believe that having it accept`&ArrayView` wasn't necessary.

Reply:
then you should call `view` here and not clone your vector? Or are `points` already views

Reply:
`pt` is an `ArrayView`, not `Vec`, so cloning is not a problem.

Reply:
I would still call `view` here so that it's clear for the reader that this is a cheap operation. (cloning is something like a "bad sign" in rust)

Reply:
I replaced the clones with `reborrow`.